### ISSUE TITLE
Rule suggestion: no variable declarations and assignments together [$50]

### ISSUE THREAD

**Issue by bedney** [eslint/eslint#2606]
I am wanting a (stylistic) rule where this would be disallowed:

```
var x = 2;
```

But this is fine:

```
var x;
x = 2;
```

In other words, variable declarations and variable assignments have to be broken into 2 pieces.

Is there already such a rule and I missed it?

Thanks!

**Comment by gyandeeps** [comment]
I dont think so if we have a rule for your use case. You can also check if there is any plugin which satisfies your use case here: https://www.npmjs.com/search?q=eslintplugin

**Comment by bedney** [comment]
I checked for a plugin and it doesn't seem like there is one for this.

**Comment by xjamundx** [comment]
There are a number of options on the [one-var rule](http://eslint.org/docs/rules/one-var) that give us part of what is wanted here. Some variation of `{ uninitialized: "always", initialized: "never" }` is sort-of what you want. It's like you're asking for `{unitialized: "always", initialized: "no-really-not-ever" }` :)

**Comment by bedney** [comment]
Yea, I guess that's true. It seems to me like the one-var rule is already getting crowded with options, though.

**Comment by xjamundx** [comment]
Agreed it's crowded, but mostly because of `let` and `const`. I think it's still the correct location for such a restriction. We'll have to wait and see what other team members think though before deciding if it's something we're able to add.

In the meantime, feel free to tweak the existing rule to meet your needs and distribute that as a plugin :)

**Comment by bedney** [comment]
I will respectfully disagree, mainly because of the name. `one-var` makes sense to me when discussing 'should there be one `var` statement to configure the locals for this function?'. But as more features get jammed into this rule, it seems to drift further and further away from the meaning imparted by that name. Maybe these other features should be split out into a rule that would be more appropriately named?

**Comment by nzakas** [comment]
This would have to be a separate rule. At the moment, new rules are very low priority as we move towards 1.0.0. Your best bet is to create the rule yourself, or else it could be a while.

**Comment by bedney** [comment]
I've put a bounty on this bug via BountySource if anyone with the necessary set of skills (which I don't have) wants to take it on.

I have no idea what we call this rule, but I'd like something that's consistent with all of the other rule naming. Maybe @nzakas or @ilyavolodin or @xjamundx have ideas here...

My acceptance criteria is that when this rule is set to 'always' (or 'never', depending on how we name the rule), these kinds of statements are disallowed:

```
var x=2, y=3, z=4;
var x=2;
var y=3;
var z=4;
```

and only these are acceptable (ignore the formatting here - I don't really care about whether things are on separate lines or whatever - that's up to the "onevar" rule or other rules - the main thing is that you cannot both declare and define all in one statement):

```
var x, y, z;

and then:

x = 2; y =3; z=4;

```

Conversely, the setting opposite to 'always' or 'never' allows this form and disallows what I describe above.

I'm open to suggestions and comments here (in fact, I'd love them!)

Thanks!

**Comment by ilyavolodin** [comment]
Should this rule error on the following code:
```js
for (var i = 0, l = array.length; i < l; i++) {
    ...
}
```
How about the following code:
```js
var myFunc = function() {
    ...
}
```

**Comment by bedney** [comment]
Most definitely yes! and yes!

**Comment by btmills** [comment]
Perhaps `initialize-declarations` with options `"always"` or `"never"`? (The `"never"` mode would of course not report `const` declarations.)

**Comment by ilyavolodin** [comment]
I would shrink the name down to `init-declarations` but I think that should work, even though it's a bit vague.

**Comment by bedney** [comment]
Another suggestion (which I'm not in love with, but I'll throw it out there):

`split-declarations-from-definitions`

**Comment by nzakas** [comment]
+1 for `init-declarations`

**Comment by bedney** [comment]
Ok, so `init-declarations` sounds good to me. Given that name, I would assume the behavior I'm looking for would be the 'never' setting. Also, I agree that 'never' should ignore 'const' declarations, but I think that 'let' should be included in the rule along with 'var'.

**Comment by Berkmann18** [comment]
Sorry for the bump.
I keep getting "Split into declaration and initialization" warnings (which I don't want) but according to this thread the rules which are init-declarations and one-var but both are off, why is that happening ?

**Comment by ilyavolodin** [comment]
@Berkmann18 Please open a new issue.
